File: ios/BriefPilot/BriefPilotScannerModule.swift
import Foundation
import React
import UIKit

@objc(BriefPilotScanner)
class BriefPilotScanner: NSObject {

  @objc static func requiresMainQueueSetup() -> Bool { false }

  // MARK: - Capabilities

  @objc(getCapabilities:rejecter:)
  func getCapabilities(
    _ resolve: RCTPromiseResolveBlock,
    rejecter _: RCTPromiseRejectBlock
  ) {
    resolve([
      "edgeDetection":         true,
      "perspectiveCorrection": true,
      "adaptiveThreshold":     true,
      "clahe":                 true,
      "shadowRemoval":         true,
      "frameProcessor":        false,
      "platform":              "ios",
      "version":               "ios-opencv-1.0",
    ])
  }

  // MARK: - Edge detection (photo)

  @objc(detectDocumentEdges:resolver:rejecter:)
  func detectDocumentEdges(
    _ payload: NSDictionary,
    resolver resolve: RCTPromiseResolveBlock,
    rejecter _: RCTPromiseRejectBlock
  ) {
    guard let uri = payload["uri"] as? String,
          let image = loadImage(from: uri) else {
      resolve(NSNull()); return
    }
    guard let result = BriefPilotOpenCVHelper.detectDocumentCorners(image) else {
      resolve(NSNull()); return
    }
    resolve(cornerToMap(result))
  }

  // MARK: - Edge detection (live frame)

  @objc(detectEdgesInFrame:resolver:rejecter:)
  func detectEdgesInFrame(
    _ payload: NSDictionary,
    resolver resolve: RCTPromiseResolveBlock,
    rejecter _: RCTPromiseRejectBlock
  ) {
    guard let uri = payload["uri"] as? String,
          let image = loadImage(from: uri) else {
      resolve(NSNull()); return
    }
    let result = BriefPilotOpenCVHelper.detectDocumentCorners(image)
    resolve(result.map { cornerToMap($0) } ?? NSNull())
  }

  // MARK: - Perspective warp

  @objc(warpPerspective:resolver:rejecter:)
  func warpPerspective(
    _ payload: NSDictionary,
    resolver resolve: RCTPromiseResolveBlock,
    rejecter reject: RCTPromiseRejectBlock
  ) {
    guard let imageUri = payload["imageUri"] as? String,
          let image = loadImage(from: imageUri) else {
      reject("INVALID_INPUT", "imageUri gerekli veya yüklenemedi", nil); return
    }

    let corners = DocumentCornerResult()
    corners.topLeft     = readPoint(payload["topLeft"])
    corners.topRight    = readPoint(payload["topRight"])
    corners.bottomRight = readPoint(payload["bottomRight"])
    corners.bottomLeft  = readPoint(payload["bottomLeft"])
    corners.confidence  = 1.0

    let outW = (payload["outWidth"]  as? Double) ?? 2480.0
    let outH = (payload["outHeight"] as? Double) ?? 3508.0

    guard let warped = BriefPilotOpenCVHelper.warpPerspective(
      image, corners: corners, outputSize: CGSize(width: outW, height: outH)
    ), let outUri = saveImage(warped) else {
      reject("WARP_FAILED", "Perspektif düzeltme başarısız", nil); return
    }
    resolve(["uri": outUri])
  }

  // MARK: - Filter

  @objc(applyFilter:resolver:rejecter:)
  func applyFilter(
    _ payload: NSDictionary,
    resolver resolve: RCTPromiseResolveBlock,
    rejecter reject: RCTPromiseRejectBlock
  ) {
    guard let imageUri = payload["imageUri"] as? String else {
      reject("INVALID_INPUT", "imageUri gerekli", nil); return
    }
    guard let image = loadImage(from: imageUri) else {
      resolve(["uri": imageUri]); return
    }
    let filter = (payload["filter"] as? String) ?? "original"

    let processed: UIImage?
    switch filter {
    case "bw", "document", "thresh":
      processed = BriefPilotOpenCVHelper.applyAdaptiveThreshold(image)
    case "enhance", "clahe":
      let clip = (payload["clipLimit"] as? Double) ?? 2.5
      processed = BriefPilotOpenCVHelper.applyCLAHE(image, clipLimit: clip)
    case "shadow":
      processed = BriefPilotOpenCVHelper.removeShadow(image)
    case "color", "colorEnhance":
      processed = BriefPilotOpenCVHelper.applyColorEnhancement(image)
    default:
      processed = image
    }

    let outUri = processed.flatMap { saveImage($0) } ?? imageUri
    resolve(["uri": outUri])
  }

  // MARK: - Magic filter

  @objc(applyMagicFilter:resolver:rejecter:)
  func applyMagicFilter(
    _ payload: NSDictionary,
    resolver resolve: RCTPromiseResolveBlock,
    rejecter reject: RCTPromiseRejectBlock
  ) {
    guard let imageUri = payload["imageUri"] as? String else {
      reject("INVALID_INPUT", "imageUri gerekli", nil); return
    }
    guard let image = loadImage(from: imageUri) else {
      resolve(["uri": imageUri]); return
    }
    let result = BriefPilotOpenCVHelper.applyDocumentMagicFilter(image)
    let outUri = result.flatMap { saveImage($0) } ?? imageUri
    resolve(["uri": outUri])
  }

  // MARK: - Helpers

  private func cornerToMap(_ r: DocumentCornerResult) -> NSDictionary {
    [
      "topLeft":        ["x": r.topLeft.x,     "y": r.topLeft.y],
      "topRight":       ["x": r.topRight.x,    "y": r.topRight.y],
      "bottomRight":    ["x": r.bottomRight.x, "y": r.bottomRight.y],
      "bottomLeft":     ["x": r.bottomLeft.x,  "y": r.bottomLeft.y],
      "confidence":     r.confidence,
      "isBlurry":       r.isBlurry,
      "needsFlash":     r.needsFlash,
      "blurVariance":   r.blurVariance,
      "avgBrightness":  r.avgBrightness,
      "areaScore":      r.areaScore,
      "angleScore":     r.angleScore,
      "aspectScore":    r.aspectScore,
      "centerScore":    r.centerScore,
    ]
  }

  private func readPoint(_ val: Any?) -> CGPoint {
    guard let d = val as? NSDictionary,
          let x = d["x"] as? Double,
          let y = d["y"] as? Double else { return .zero }
    return CGPoint(x: x, y: y)
  }

  private func loadImage(from uri: String) -> UIImage? {
    let path = uri.hasPrefix("file://") ? String(uri.dropFirst(7)) : uri
    return UIImage(contentsOfFile: path)
  }

  private func saveImage(_ image: UIImage) -> String? {
    guard let data = image.jpegData(compressionQuality: 0.92) else { return nil }
    let name = "bp_scan_\(Int(Date().timeIntervalSince1970 * 1000)).jpg"
    let url  = URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent(name)
    do {
      try data.write(to: url)
      return url.absoluteString
    } catch { return nil }
  }
}

File: android/app/src/main/java/com/briefpilot/app/scanner/BriefPilotScannerModule.kt
package com.briefpilot.app.scanner

import com.facebook.react.bridge.*
import org.opencv.android.OpenCVLoader

class BriefPilotScannerModule(private val reactContext: ReactApplicationContext)
    : ReactContextBaseJavaModule(reactContext) {

    override fun getName() = "BriefPilotScanner"

    override fun initialize() {
        super.initialize()
        OpenCVLoader.initLocal() // OpenCV'yi başlat
    }

    @ReactMethod
    fun getCapabilities(promise: Promise) {
        val map = WritableNativeMap().apply {
            putBoolean("edgeDetection",         true)
            putBoolean("perspectiveCorrection",  true)
            putBoolean("adaptiveThreshold",      true)
            putBoolean("clahe",                  true)
            putBoolean("shadowRemoval",          false)
            putBoolean("frameProcessor",         false)
            putString("platform",  "android")
            putString("version",   "android-opencv-1.0")
        }
        promise.resolve(map)
    }

    @ReactMethod
    fun detectDocumentEdges(payload: ReadableMap, promise: Promise) {
        val uri = payload.getString("uri") ?: run { promise.resolve(null); return }
        val bitmap = BriefPilotOpenCVHelper.loadBitmap(uri) ?: run { promise.resolve(null); return }

        Thread {
            try {
                val result = BriefPilotOpenCVHelper.detectDocumentCorners(bitmap)
                if (result == null) { promise.resolve(null); return@Thread }
                promise.resolve(cornerToMap(result))
            } catch (e: Exception) {
                promise.reject("DETECT_FAILED", e.message, e)
            }
        }.start()
    }

    @ReactMethod
    fun detectEdgesInFrame(payload: ReadableMap, promise: Promise) {
        val uri = payload.getString("uri") ?: run { promise.resolve(null); return }
        val bitmap = BriefPilotOpenCVHelper.loadBitmap(uri) ?: run { promise.resolve(null); return }

        Thread {
            try {
                val result = BriefPilotOpenCVHelper.detectDocumentCorners(bitmap)
                promise.resolve(if (result != null) cornerToMap(result) else null)
            } catch (e: Exception) {
                promise.resolve(null)
            }
        }.start()
    }

    @ReactMethod
    fun warpPerspective(payload: ReadableMap, promise: Promise) {
        val uri = payload.getString("imageUri") ?: run { promise.reject("INVALID_INPUT", "imageUri gerekli"); return }
        val bitmap = BriefPilotOpenCVHelper.loadBitmap(uri) ?: run { promise.reject("LOAD_FAILED", "Görüntü yüklenemedi"); return }

        val corners = BriefPilotOpenCVHelper.CornerResult(
            topLeft     = readPointF(payload.getMap("topLeft")),
            topRight    = readPointF(payload.getMap("topRight")),
            bottomRight = readPointF(payload.getMap("bottomRight")),
            bottomLeft  = readPointF(payload.getMap("bottomLeft")),
            confidence  = 1f,
            isBlurry    = false, needsFlash = false,
            blurVariance = 0.0, avgBrightness = 0.0,
        )

        val outW = if (payload.hasKey("outWidth"))  payload.getInt("outWidth")  else 2480
        val outH = if (payload.hasKey("outHeight")) payload.getInt("outHeight") else 3508

        Thread {
            try {
                val warped = BriefPilotOpenCVHelper.warpPerspective(bitmap, corners, outW, outH)
                val outUri = warped?.let { BriefPilotOpenCVHelper.saveBitmap(it, reactContext.cacheDir) }
                if (outUri != null) {
                    promise.resolve(WritableNativeMap().apply { putString("uri", outUri) })
                } else {
                    promise.reject("WARP_FAILED", "Perspektif düzeltme başarısız")
                }
            } catch (e: Exception) {
                promise.reject("WARP_FAILED", e.message, e)
            }
        }.start()
    }

    @ReactMethod
    fun applyFilter(payload: ReadableMap, promise: Promise) {
        val uri    = payload.getString("imageUri") ?: run { promise.reject("INVALID_INPUT", "imageUri gerekli"); return }
        val filter = payload.getString("filter") ?: "original"
        val bitmap = BriefPilotOpenCVHelper.loadBitmap(uri) ?: run { promise.resolve(WritableNativeMap().apply { putString("uri", uri) }); return }

        Thread {
            try {
                val processed = when (filter) {
                    "bw", "document", "thresh" -> BriefPilotOpenCVHelper.applyAdaptiveThreshold(bitmap)
                    "enhance", "clahe"         -> {
                        val clip = if (payload.hasKey("clipLimit")) payload.getDouble("clipLimit") else 2.0
                        BriefPilotOpenCVHelper.applyCLAHE(bitmap, clip)
                    }
                    else -> bitmap
                }
                val outUri = processed?.let { BriefPilotOpenCVHelper.saveBitmap(it, reactContext.cacheDir) } ?: uri
                promise.resolve(WritableNativeMap().apply { putString("uri", outUri) })
            } catch (e: Exception) {
                promise.resolve(WritableNativeMap().apply { putString("uri", uri) })
            }
        }.start()
    }

    @ReactMethod
    fun applyMagicFilter(payload: ReadableMap, promise: Promise) {
        val uri = payload.getString("imageUri") ?: run { promise.reject("INVALID_INPUT", "imageUri gerekli"); return }
        val bitmap = BriefPilotOpenCVHelper.loadBitmap(uri) ?: run { promise.resolve(WritableNativeMap().apply { putString("uri", uri) }); return }

        Thread {
            try {
                val result = BriefPilotOpenCVHelper.applyDocumentMagicFilter(bitmap)
                val outUri = result?.let { BriefPilotOpenCVHelper.saveBitmap(it, reactContext.cacheDir) } ?: uri
                promise.resolve(WritableNativeMap().apply { putString("uri", outUri) })
            } catch (e: Exception) {
                promise.resolve(WritableNativeMap().apply { putString("uri", uri) })
            }
        }.start()
    }

    // ── Yardımcılar ─────────────────────────────────────────────────────────

    private fun cornerToMap(r: BriefPilotOpenCVHelper.CornerResult): WritableMap {
        return WritableNativeMap().apply {
            putDouble("confidence",    r.confidence.toDouble())
            putBoolean("isBlurry",     r.isBlurry)
            putBoolean("needsFlash",   r.needsFlash)
            putDouble("blurVariance",  r.blurVariance)
            putDouble("avgBrightness", r.avgBrightness)
            if (r.confidence > 0) {
                putMap("topLeft",     pointMap(r.topLeft))
                putMap("topRight",    pointMap(r.topRight))
                putMap("bottomRight", pointMap(r.bottomRight))
                putMap("bottomLeft",  pointMap(r.bottomLeft))
            }
        }
    }

    private fun pointMap(p: BriefPilotOpenCVHelper.PointF) = WritableNativeMap().apply {
        putDouble("x", p.x.toDouble())
        putDouble("y", p.y.toDouble())
    }

    private fun readPointF(map: ReadableMap?): BriefPilotOpenCVHelper.PointF {
        if (map == null) return BriefPilotOpenCVHelper.PointF(0f, 0f)
        return BriefPilotOpenCVHelper.PointF(
            map.getDouble("x").toFloat(),
            map.getDouble("y").toFloat(),
        )
    }
}
